### ISSUE TITLE
Multi-Threading - LOW-RAM-Mode - Single Thread only?

### ISSUE THREAD

**Issue by JamesPous** [JacobDev1/xl-converter#53]
### Version

1.0.2

### OS

Windows

### Description

Thank you for your further development of XL-Converter and improvements.
The advanced settings are fantastic to reduce damaged destination files and prevent larger files than the orignal!

The latest Multi-Threading Performance Mode use all my selected threads and the RAM usage is good.
But If I switch to LOW-RAM mode, I will not see any big difference in RAM usage. It will only use one thread for cjpegli.exe! Is it your intension? I expected, that xl-converter use all selected threads for low ram mode, too.

### Steps to Reproduce

Activate low ram mode

jpeg to jpxl recompression - effort 9

### Expected Result

using all threads

### Current Result

single thread only!

### Input Formats

JPEG

### Output Format

JPEG recompression

**Comment by JacobDev1** [comment]
Hello again, thanks staying interested in my program.

Let's just say I spoiled my users with parallel encoding so much, that regular encoding feels like such a downgrade. In your case, the JPEG XL encoder receives all threads but is slower because images aren't fetched from disk simultaneously.

I'm happy to explain how this works.
|Mode|Explanation|Example|
|-|-|-|
|Performance|Encoders are run in parallel. This results in multiple copies and much higher RAM usage.|For 8 threads, up to 8 encoders would be spawned.|
|Low RAM|Encoders are run sequentially. This results in one encoder running at a time with all threads passed to it. Recommended only for big images.|For 8 threads, only 1 encoder would be spawned with 8 threads made available to it.|

Some encoders do a good job and some don't.

The official reference JPEGLI encoder is single-threaded because there is no way to pass threads to it. It just... is like that.

You don't see any difference in RAM usage because "Lossless JPEG Transcoding" requires a tiny amount of resources. I checked, and all threads are passed correctly to it. I do agree, it's not very fast. The I/O bottleneck is slashing the performance quite hard here. That is how fast encoders run without XL Converter accelerating them.

The "Low RAM" mode is half-baked, because it was added at the last minute before an update as someone complained about high RAM usage (#49). In short, they used "Effort" 9 and 10 with JPEG XL, their data set had some very high resolution images, and their computer ran out of RAM. The "Low RAM" mode is meant only for processing extremely large images with higher "Efforts".

Don't worry too much about it. Set it to "Performance" and forget it. It is for those rare cases where RAM is an issue. This function will change, I'm still working out the details.

I'm pleased to hear you have no issue with high RAM usage. Discussing it in #49 had me worried. Thanks for the report.

---

Oh, and if you're wondering why I call "Lossless JPEG Recompression" "Lossless JPEG Transcoding", I'm renaming it in `v1.0.3` as per this [Arxiv paper](https://arxiv.org/abs/2208.11673). I named it after some related line found in the `cjxl` encoder but only now I managed to stumble on this paper.

**Comment by JamesPous** [comment]
Thank you for the detailed explanation of the two modes (Performance/Low RAM).

I'm not a fan of mode names that don't directly or functionally describe a switch (e.g. slow, medium, fast).
But it's fine in XL Convert because the field help texts are displayed when you move the mouse. I have just noticed this. I think this is very well solved in the new version. I might have named the function differently. But that's a matter of taste.

I can understand, that other persons need this setting to convert large picture files and prevent out of ram errors.
It is great that your tool can handle it. But for jpeg transcoding it works fine and the RAM usage is smaller!

I hope you support JPEGXL and JPEGLI in the future, because the developer split it into two projects with the next version 0.11.

You can close this thread!

**Comment by JacobDev1** [comment]
I'm not a fan of this naming convention either. The original names I considered were "Run Encoders in" - "Parallel" and "Sequence" but I was worried users wouldn't understand it. I didn't want to delay the update just because of the naming.

I plan to change this option the next update. The "Low RAM" mode exists only to cover an edge case.

Yes, I noticed the project split. I'm still waiting for any tagged releases.
